src/arq/_saw.py

```diff
@@ -0,0 +1,133 @@
+# -*- coding: utf-8 -*-
+
+
+import gevent.queue
+import itertools
+import random
+import struct
+
+from binascii import hexlify
+from enum import IntEnum
+
+from ._utils import concurrently
+
+
+class PacketType(IntEnum):
+    data = 0
+    ackn = 1
+
+
+HEADER = struct.Struct('>BH')
+HEADER_SIZE = HEADER.size
+
+
+def sequence_numbers():
+    """Generate sequence numbers that wrap around."""
+    for i in itertools.count(start=1):
+        yield i % 0xffff
+
+
+def recvfrom(socket, peer, size):
+    """Receive a packet from a designated peer."""
+    data, p = socket.recvfrom(size)
+    while p != peer:
+        data, p = socket.recvfrom(size)
+    return data
+
+
+def stop_and_wait(socket, peer, iq, oq, prefix='', p=0.0):
+    """."""
```

Comment:
**TODO**: explain that `iq` and `oq` should be bounded to implement back-pressure.

Reply:
Done.